GORA-532: Apache Gora Benchmark Module

sneceesay77 opened the pull request:
Initial pull request.

lewismc commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkClient.java:
```diff
@@ -0,0 +1,273 @@
+       
+        
+    } catch (GoraException e) {
+      //e.printStackTrace();
```
Remove commented out code

lewismc commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkClient.java:
```diff
@@ -0,0 +1,273 @@
+  @Override
+  public int delete(String table, String key) {
+    try {
+      //LOG.info("Deleting Key "+key+" From Table "+table);
```
Delete

lewismc commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkClient.java:
```diff
@@ -0,0 +1,273 @@
+   */
+  @Override
+  public int insert(String table, String key, HashMap<String, ByteIterator> values) {
+    // TODO Auto-generated method stub
```
Always remove auto-generated stubs

lewismc commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkClient.java:
```diff
@@ -0,0 +1,273 @@
+      }
+      dataStore.put(key, user);
+    } catch (Exception e) {
+      //e.printStackTrace();
```
Remove

lewismc commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkClient.java:
```diff
@@ -0,0 +1,273 @@
+      List<String> allFields = Arrays.asList(FIELDS);
+      for(String field: values.keySet()) {
+        if(GoraBenchmarkUtils.isFieldUpdatable(field, values)) {
+          /**Get the index of the field from the global fields array and call the right method*/
```
Make comment ```//```

lewismc commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkUtils.java:
```diff
@@ -0,0 +1,453 @@
+package org.apache.gora.benchmark;
```
License header

lewismc commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkUtils.java:
```diff
@@ -0,0 +1,453 @@
+
+// TODO: Auto-generated Javadoc
+/**
+ * The Class GoraBenchmarkUtils.
```
Please describe what the class does and how it is run. What input is expected and what output is expected.

sneceesay77 replied:
Updated the file with some description, the methods in this file are called in the init method of GoraBenchmarkClient. The class is not directly executed. Basically, you can specify how many fields a particular table in a particular datastore should have.

lewismc replied:
Thanks for updating with a description.

lewismc commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkUtils.java:
```diff
@@ -0,0 +1,453 @@
+  
+  /** The Constant DEFAULT_DATA_STORE_KEY. */
+  private static final String DEFAULT_DATA_STORE_KEY = "gora.datastore.default";
+  
```
Remove whitespace

lewismc commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkUtils.java:
```diff
@@ -0,0 +1,453 @@
+  
+  private static final Logger LOG = LoggerFactory.getLogger(GoraBenchmarkUtils.class);
+  
+  
```
Remove whitespace

lewismc commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkUtils.java:
```diff
@@ -0,0 +1,453 @@
+      avroWriter.write(avroSchema.toString(4));
+      avroWriter.close();
+    } catch (FileNotFoundException e) {
+      //e.printStackTrace();
```
Remove comment

lewismc commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkUtils.java:
```diff
@@ -0,0 +1,453 @@
+      //e.printStackTrace();
+    } catch (IOException e) {
+      // TODO Auto-generated catch block
+      //e.printStackTrace();
```
Remove comment

lewismc commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkUtils.java:
```diff
@@ -0,0 +1,453 @@
+
+      Element rootNode = mappingDocument.createElement(GORA_ROOT_ELEMENT);
+      mappingDocument.appendChild(rootNode);
+      //System.out.println(AVRO_FULL_PATH);
```
Remove comment

lewismc commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkUtils.java:
```diff
@@ -0,0 +1,453 @@
+      StreamResult consoleResult = new StreamResult(new StringWriter());
+      transformer.transform(source, consoleResult);
+      String xmlString = consoleResult.getWriter().toString();
+      //System.out.println(xmlString);
```
Remove comment

lewismc commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkUtils.java:
```diff
@@ -0,0 +1,453 @@
+      //System.out.println(xmlString);
+    } catch (ParserConfigurationException e) {
+      // TODO Auto-generated catch block
+      //e.printStackTrace();
```
Remove comment

lewismc commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkUtils.java:
```diff
@@ -0,0 +1,453 @@
+      //e.printStackTrace();
+    } catch (TransformerConfigurationException e) {
+      // TODO Auto-generated catch block
+      //e.printStackTrace();
```
Remove comment

lewismc commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkUtils.java:
```diff
@@ -0,0 +1,453 @@
+      //e.printStackTrace();
+    } catch (TransformerException e) {
+      // TODO Auto-generated catch blockString
+      //e.printStackTrace();DEFAULT_DATA_STORE_KEY
```
Remove comment

lewismc commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkUtils.java:
```diff
@@ -0,0 +1,453 @@
+      GoraCompiler.compileSchema(inputFiles, BEAN_DESTINATION_DIR, licenseHeader);
+    } catch (IOException e) {
+      // TODO Auto-generated catch block
+      //e.printStackTrace();
```
Remove comments

lewismc commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkUtils.java:
```diff
@@ -0,0 +1,453 @@
+      jsonObject = new JSONObject(jsonString);
+    } catch (IOException e) {
+      // TODO Auto-generated catch block
+      //e.printStackTrace();
```
Remove comments

lewismc commented on gora-benchmark/src/main/resources/gora-couchdb-mapping.xml:
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
License header

sneceesay77 replied:
Mapping files are auto-generated. I have the code to add the license file accordingly.

lewismc commented on gora-benchmark/src/main/resources/gora-hbase-mapping.xml:
```diff
@@ -0,0 +1,59 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
License header

sneceesay77 replied:
Mapping files are auto-generated. I have the code to add the license file accordingly.

lewismc commented on gora-benchmark/src/main/resources/gora-mongodb-mapping.xml:
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
License header

sneceesay77 replied:
Mapping files are auto-generated. I have added the code to add the license file accordingly.

lewismc commented on gora-benchmark/src/test/java/org/apache/gora/benchmark/GoraClientTest.java:
```diff
@@ -0,0 +1,201 @@
+package org.apache.gora.benchmark;
```
License header

lewismc commented on gora-benchmark/src/test/java/org/apache/gora/benchmark/GoraClientTest.java:
```diff
@@ -0,0 +1,201 @@
+
+import generated.User;
+
+// TODO: Auto-generated Javadoc
```
Remove comment

lewismc commented on gora-benchmark/src/test/java/org/apache/gora/benchmark/GoraClientTest.java:
```diff
@@ -0,0 +1,201 @@
+    //dataStores.add("mongodb");
+    //dataStores.add("hbase");
+    //bmutils.generateAvroSchema(NUMBER_OF_FIELDS);
+    //for(String dataStore: dataStores) {
```
Remove all commented-out code

lewismc commented on gora-benchmark/src/test/java/org/apache/gora/benchmark/GoraClientTest.java:
```diff
@@ -0,0 +1,201 @@
+  public void testScan(){
+    Vector<HashMap<String, ByteIterator>> results = new Vector<HashMap<String, ByteIterator>>();
+    Set<String> fields = new HashSet<>();
+    //fields.add("field0");
```
Remove

lewismc commented on gora-benchmark/src/test/java/org/apache/gora/benchmark/GoraClientTest.java:
```diff
@@ -0,0 +1,201 @@
+     User u = readRecord("key1");
+     assertEquals("updated0", u.getField0().toString());
+    }
+    //Read Record from 
```
Remove

lewismc commented on gora-benchmark/workloads/tsworkload_template:
```diff
@@ -0,0 +1,283 @@
+# Copyright (c) 2017 YCSB contributors. All rights reserved.
```
Remove gora-benchmark/workloads/.workloadb.swp above this should not be available within the pull request

sneceesay77 commented on gora-benchmark/pom.xml:
```diff
@@ -0,0 +1,197 @@
+			<groupId>org.apache.gora</groupId>
+			<artifactId>gora-mongodb</artifactId>
+			<scope>compile</scope>
+			<version>0.9-SNAPSHOT </version>
```
Like others, there is no definition or declaration of MongoDB in the parent pom file, hence the explicit declaration for now. Maybe I can create a separate issue for this or should I just go ahead and edit the parent pom accordingly.

lewismc replied:
+1, just go ahead and obtain this from parent pom. Do whatever you need to do for that.

sneceesay77 commented on gora-benchmark/src/main/avro/person.json:
```diff
@@ -0,0 +1,18 @@
+  "name": "Person", "default":null,
+  "namespace": "generated",
+  "fields" : [
+    {"name": "userId", "type": ["null","string"], "default":null},
```
I am not actually using this file anymore, but yes it would be interesting to test with complex objects. For YCSB it uses a simple User object which can have **n** number of fields.

lewismc replied:
OK no problem. Please remote file. Thanks

sneceesay77 commented on gora-benchmark/src/main/avro/user.json:
```diff
@@ -0,0 +1,112 @@
+    "namespace": "generated",
+    "type": "record",
+    "fields": [
+        {
```
The autogenerated JSON files are all the same.  The person.json file is not being used. I will remove the file.

sneceesay77 commented on gora-benchmark/runWorkloads.sh:
```diff
@@ -0,0 +1,119 @@
+     ./gora-bench.sh run -threads 15 -s -p readallfields=true -p operationcount=$i -p measurementtype=timeseries -p timeseries.granularity=2000 -P workloads/$3 | tee -a $outputfile
+   elif [ $2 = "ycsb" ]
+   then
+      #../../ycsb/bin/ycsb.sh load mongodb -threads 15 -s -p fieldcount=20 -p recordcount=$insertfactor -p operationcount=$i -P workloads/workloada
```
This is an Adhoc script to automate the running of experiments, I don't think it should be included. The readme file will cover essential parts of this code.

lewismc replied:
Got it. Please remove it if not needed thanks.

lewismc commented:
Please ping me here @sneceesay77 when the PR updated and you've provided a README of how we can run this. Thank you

sneceesay77 commented:
> Please ping me here @sneceesay77 when the PR updated and you've provided a README of how we can run this. Thank you

sneceesay77 commented:
> Please ping me here @sneceesay77 when the PR updated and you've provided a README of how we can run this. Thank you

@lewismc, were you able to look at this for me? I have completed all the requested changes.

lewismc commented on gora-benchmark/README.md:
```diff
@@ -0,0 +1,48 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+This guidline outlines the process of compiling and running Gora Benchmark Module
```
guidline --> guideline

lewismc commented on gora-benchmark/README.md:
```diff
@@ -0,0 +1,48 @@
+
+2. From the module directory i.e. gora-benchmark, run mvn clean install
+3. This should run some test and successfully build the module. 
+	If the build fails and complains about maven not able to ycsb-core-verion.jar then download the jar manually and add it to ~/.m2/repository/com/yahoo/ycsb/core/0.1.4/
```
Why would this ever be the case? Can you elaborate?

lewismc replied:
I experienced exactly this issue
```
[ERROR] Failed to execute goal on project gora-benchmark: Could not resolve dependencies for project org.apache.gora:gora-benchmark:jar:0.9-SNAPSHOT: Could not find artifact com.yahoo.ycsb:core:jar:0.1.4 in central (https://repo.maven.apache.org/maven2) -> [Help 1]
[ERROR]
```

sneceesay77 replied:
We are now working with YCSB to resolve this issue.

lewismc commented on gora-benchmark/pom.xml:
```diff
@@ -0,0 +1,188 @@
+      <groupId>org.apache.gora</groupId>
+      <artifactId>gora-mongodb</artifactId>
+      <scope>compile</scope>
+      <version>0.9-SNAPSHOT </version>
```
This issue is still not resolved. Can you please update it otherwise this module will break when we release another version. Thanks

lewismc commented on gora-benchmark/pom.xml:
```diff
@@ -0,0 +1,188 @@
+      <groupId>com.yahoo.ycsb</groupId>
+      <artifactId>core</artifactId>
+      <version>0.1.4</version>
+    </dependency>
```
If this dependency is not available from Maven central then we need to address that. How are you currently fetching it?

sneceesay77 replied:
I understand this isn't the standard approach but what I did was to download the jar file manually from maven website and place it in ~/.m2/repository/com/yahoo/ycsb/core/0.1.4/. I saw this trick from [1]. Also [2] and [3] experienced similar problems. 

I will log an issue on YCSB GitHub page. 

[1] https://stackoverflow.com/questions/22600408/use-maven-to-compile-ycsb
[2] https://github.com/blockchain-jd-com/jdchain/issues/13
[3] https://github.com/brianfrankcooper/YCSB/issues/174

sneceesay77 replied:
I have got some feedback from YCSB maintainers [1]. So basically, they have not published ycsb in maven central.

[1] https://github.com/brianfrankcooper/YCSB/issues/1340

lewismc replied:
We need to help them on that issue. I will work with them to do that. Hopefully when they release 0.17.0 we can upgrade here.

djkevincr commented on gora-tutorial/conf/gora.properties:
```diff
@@ -66,3 +66,18 @@ gora.datastore.jcache.hazelcast.config=hazelcast.xml
+# MongoDBStore properties  #
+############################
+
+gora.datastore.autocreateschema=true
```
Please take a update/merge from current master, one of your previous PR's addressed this. Properties gora.datastore.autocreateschema, gora.datastore.default are duplicated keys for in properties file. We need to avoid that.

sneceesay77 replied:
Not sure how this happened. I have already merged master with my branch thinking that all should be okay.  I will go ahead and remove the duplicate lines.

sneceesay77 replied:
I have updated my branch with the current master.  This is now resolved.

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkClient.java:
```diff
@@ -0,0 +1,266 @@
+        goraBenchmarkUtils.generateMappingFile(dataStoreName);
+        goraBenchmarkUtils.generateDataBeans();
+      }
+    } catch (GoraException e) {
```
It seems we have omitted handling the exception. May be we should at least log the exception trace.

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkUtils.java:
```diff
@@ -0,0 +1,426 @@
+      avroWriter.write(avroSchema.toString(4));
+      avroWriter.close();
+    } catch (FileNotFoundException e) {
+    } catch (IOException e) {
```
It seems we have omitted handling the exception. May be we should at least log the exception trace.

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkUtils.java:
```diff
@@ -0,0 +1,426 @@
+      Document mappingDocument = docBuilder.newDocument();
+      Element rootNode = mappingDocument.createElement(GORA_ROOT_ELEMENT);
+       
+      
```
Shall we remove these unnecessary new lines and remove commented code?

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkUtils.java:
```diff
@@ -0,0 +1,426 @@
+      StreamResult consoleResult = new StreamResult(new StringWriter());
+      transformer.transform(source, consoleResult);
+    } catch (ParserConfigurationException e) {
+    } catch (TransformerConfigurationException e) {
```
It seems we have omitted handling the exception. May be we should at least log the exception trace.

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkUtils.java:
```diff
@@ -0,0 +1,426 @@
+    File[] inputFiles = getInputFiles();
+    try {
+      GoraCompiler.compileSchema(inputFiles, BEAN_DESTINATION_DIR, licenseHeader);
+    } catch (IOException e) {
```
It seems we have omitted handling the exception. May be we should at least log the exception trace.

sneceesay77 commented:
@lewismc, so what's the way forward regarding the YCSB dependency issue? I have updated my local dependency to the latest version of YCSB i.e. 0.15.0 and updated my code accordingly. I have added these changes to this pull request. 

Right now the simplest way to go around this without maven is to:

1. Download YCSB from [1] and copy all the jar files in `ycsb-0.15.0/lib` directory to `gora-benchmark/lib`. I have tested it and it is working with both MongoDB and HBase.
2. Copy `ycsb-0.15.0/lib/core-0.15.0.jar` to ~/.m2/repository/com/yahoo/ycsb/core/0.15.0/
3. mvn clean install

[1] https://github.com/brianfrankcooper/YCSB/releases/download/0.15.0/ycsb-0.15.0.tar.gz

@djkevincr, I have updated your comments in the latest commit.

djkevincr commented on gora-benchmark/pom.xml:
```diff
@@ -0,0 +1,186 @@
+    <dependency>
+      <groupId>com.yahoo.ycsb</groupId>
+      <artifactId>core</artifactId>
+      <version>0.15.0</version>
```
I think you should define these version in parent pom, define YCSB dependency parent pom it self  and inherit UCSB dependency for module gora-benchmark. Currently this is hardcoded to benchmark module. Please check approach on other modules. Once the YCSB artifacts are available from maven central, you can address this additionally.

sneceesay77 replied:
Okay, once the maven central issue is addressed then I will handle this accordingly. Does that sounds good?

lewismc replied:
Yes

djkevincr commented on gora-benchmark/gora-bench.sh:
```diff
@@ -0,0 +1,75 @@
+# limitations under the License.
+
+#Adapted from YCSB's version of ycsb.sh. 
+CLASSPATH="."
```
What's your thoughts on moving gora-bench.sh script to GORA_HOME/bin, this is where we keep all the executable scripts. One possible option is to merge with current gora.sh, however think keeping it separately is much cleaner approach. Let us know your thoughts.

sneceesay77 replied:
Like you said, I think for now it is best to keep them separate. I will move the script to GORA_HOME/bin.

djkevincr commented on gora-benchmark/pom.xml:
```diff
@@ -0,0 +1,186 @@
+    <artifactId>gora</artifactId>
+    <version>0.9-SNAPSHOT</version>
+  </parent>
+  <artifactId>gora-benchmark</artifactId>
```
Currently you have not added the new benchmark module to parent pom under modules section, that means this module is omitted when entire project is built from root parent pom.
EG:-
```
<modules>
    <module>gora-compiler</module>
    <module>gora-compiler-cli</module>
    <module>gora-core</module>
    <module>gora-pig</module>
    <module>gora-accumulo</module>
    <module>gora-cassandra</module>
    <module>gora-goraci</module>
    <module>gora-hbase</module>
    <module>gora-infinispan</module>
    <module>gora-jcache</module>
    <module>gora-orientdb</module>
    <module>gora-lucene</module>
    <module>gora-dynamodb</module>
    <module>gora-couchdb</module>
    <module>gora-maven-plugin</module>
    <module>gora-mongodb</module>
    <module>gora-solr</module>
    <module>gora-aerospike</module>
    <module>gora-ignite</module>
    <module>gora-tutorial</module>
    <module>gora-benchmark</module>
    <module>sources-dist</module>
  </modules>
```
I think we once we have these YCSB artifacts from maven central, we can enable benchmark module to build process.

lewismc replied:
I agree

djkevincr commented on gora-benchmark/gora-workloada.csv:
```diff
@@ -0,0 +1,251 @@
+records,timems,recordsk,timesec,platform
```
May be we should consider moving these csv resources to single folder?

sneceesay77 replied:
I think will be best to not include the CSV files at all. So I have excluded all of them.

djkevincr commented on gora-benchmark/pom.xml:
```diff
@@ -0,0 +1,186 @@
+    <dependency>
+      <groupId>org.fluttercode.datafactory</groupId>
+      <artifactId>datafactory</artifactId>
+      <version>0.8</version>
```
Please define the dependencies and versions in parent pom.

sneceesay77 replied:
Done

djkevincr commented on gora-benchmark/pom.xml:
```diff
@@ -0,0 +1,186 @@
+    <dependency>
+      <groupId>org.ektorp</groupId>
+      <artifactId>org.ektorp</artifactId>
+      <version>1.4.2</version>
```
Please define the dependencies and versions in parent pom. In this case use CoughDB client version already defined in the parent pom.

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkClient.java:
```diff
@@ -0,0 +1,271 @@
+  private static final Logger LOG = LoggerFactory.getLogger(GoraBenchmarkClient.class);
+  private static final String FIELDS[] = User._ALL_FIELDS;
+  private static volatile boolean executed;
+  public static int fieldCount;
```
This field can be private access.

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkClient.java:
```diff
@@ -0,0 +1,271 @@
+  /** This is only for set to array conversion in {@link read()} method */
+  private String[] DUMMY_ARRAY = new String[0];
+  DataStore<String, User> dataStore;
+  GoraBenchmarkUtils goraBenchmarkUtils = new GoraBenchmarkUtils();
```
Please specify access modifier for this field EG:- private

djkevincr replied:
Usually we don't instantiate utility classes. Utility classes should only have set of static class level methods, you can call without instantiating. 
Eg:-  GoraBenchmarkUtils.generateMappingFile() 
One attribute of a utility classes is that, it usually never maintain state inside the class.
I think we should consider renaming this class appropriately.

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkClient.java:
```diff
@@ -0,0 +1,271 @@
+  private String[] DUMMY_ARRAY = new String[0];
+  DataStore<String, User> dataStore;
+  GoraBenchmarkUtils goraBenchmarkUtils = new GoraBenchmarkUtils();
+  User user = new User();
```
Please specify access modifier for this field EG:- private

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkClient.java:
```diff
@@ -0,0 +1,271 @@
+      String persistentClass = prop.getProperty("persistent.class", "org.apache.gora.benchmark.generated.User");
+      Properties p = DataStoreFactory.createProps();
+      dataStore = DataStoreFactory.getDataStore(keyClass, persistentClass, p, new Configuration());
+      synchronized (GoraBenchmarkClient.class) {
```
I noticed class level synchronization is used here as opposed object level.
class level - synchronized (GoraBenchmarkClient.class) { }
object level - synchronized (this) { }
Can you please explain, why you have chosen class level locking as opposed to object level?

sneceesay77 replied:
@djkevincr, thanks for the comment. I am not entirely sure if this is the best approach, but my choice of using class level synchronisation is to avoid any form of interference from other objects. According to my understanding of how Gora works, the resources I am creating (mapping files, data beans and avro schemas) in the synchronization block should be created once. So I thought the safest way of doing this would be to use class level synchronization.  Also, I am using a shared variable (private static volatile boolean executed) in the synchronized block that checks whether the code block has been executed by other threads. 

As for the second part, of you question, can you please elaborate further?

djkevincr replied:
So basically, when class level locking is used, for multiple instances of the same GoraBenchmarkClient class in single JVM, only one thread will get access to class level monitor ( only a single monitor per class of instances )  and execute synchronized block of a single instance from all the instances available in JVM. In object level synchronization, threads just have to wait for each instance's monitor as opposed to class object monitor. Threads can paralleled acquire instance monitors. 

Class level synchronization can lead to higher thread contention and possibly higher waiting times for threads. This can be a overhead since if we don't really want to synchronize on all instances of the class and synchronization is only required for per single instance. If you compare two case class level and object level, class level has the highest level synchronization/safety as mentioned but it comes with overhead. 

Think about two client instances of GoraBenchmarkClient and how these instances are getting accessed/methods being call when a we perform a benchmark. Think about above statements can have a impact. These kind of concurrency related things might influence benchmark results we get.

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkClient.java:
```diff
@@ -0,0 +1,271 @@
+        goraBenchmarkUtils.generateDataBeans();
+      }
+    } catch (GoraException e) {
+      LOG.info("There is a problem in initialising the DataStore \n"+e.getMessage());
```
This code is not formatted correctly. Use always parameterized logging if you want to parameterize the message content. Also we should not omit the exception trace added to to logging.
Eg:- 
LOG.info("There is a problem in initialising the DataStore \n {}", e.getMessage(), e);

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkClient.java:
```diff
@@ -0,0 +1,271 @@
+  public static int fieldCount;
+  /** This is only for set to array conversion in {@link read()} method */
+  private String[] DUMMY_ARRAY = new String[0];
+  DataStore<String, User> dataStore;
```
Please specify access modifier for this field EG:- private

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkClient.java:
```diff
@@ -0,0 +1,271 @@
+
+  /***
+   * Initialisation method. This method is called once for each database
+   * instance.
```
Please document the exception.

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkClient.java:
```diff
@@ -0,0 +1,271 @@
+   * Cleanup any state for this DB.
+   * 
+   * It is very important to close the datastore properly, otherwise some data
+   * loss might occur.
```
Please document the exception.

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkClient.java:
```diff
@@ -0,0 +1,271 @@
+          .parseInt(prop.getProperty(CoreWorkload.FIELD_COUNT_PROPERTY, CoreWorkload.FIELD_COUNT_PROPERTY_DEFAULT));
+      String keyClass = prop.getProperty("key.class", "java.lang.String");
+      String persistentClass = prop.getProperty("persistent.class", "org.apache.gora.benchmark.generated.User");
+      Properties p = DataStoreFactory.createProps();
```
p - Please use proper variable names.

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkClient.java:
```diff
@@ -0,0 +1,271 @@
+      String keyClass = prop.getProperty("key.class", "java.lang.String");
+      String persistentClass = prop.getProperty("persistent.class", "org.apache.gora.benchmark.generated.User");
+      Properties p = DataStoreFactory.createProps();
+      dataStore = DataStoreFactory.getDataStore(keyClass, persistentClass, p, new Configuration());
```
Here dataStore is created outside the below synchronization block. Is there a possibility of creating multiple untracked dataStore instances at concurrency? 
dataStore = DataStoreFactory.getDataStore(keyClass, persistentClass, p, new Configuration());

sneceesay77 replied:
@djkevincr, yes that should be possible. Is there a reason for this? What I know is, when running the benchmark client the number of threads to start can be supplied and the implemented init() method of YCSB framework is executed once by each of those threads. Maybe if you explain more then I can provide a better explanation.

djkevincr replied:
My point is there s no need to create a datastore instance as above for a already initialized GoraBenchmarkClient  instance. Let's say one thread executes init() method and set executed boolean variable value to 'true', generate mapping file etc and returns. Lets say a second and third thread come and execute executes init(), it checks executed variable since it s true it will return without doing doing generation of mapping file etc. For second and third threads, there s overhead to creating that datastore instance and override the class level field. 1st, 2nd created will be garbage collected, when you override that class level field from second and third threads. Which is I think unnecessary. 
If you move that DataStoreFactory.getDataStore into synchronize block, there will be no additional datastore object creation and will create only one instance. This is true for all the class level field value you set before the synchronize block. Those are overhead. Datastore creation is pretty expensive operation due it s network call etc nature.

sneceesay77 replied:
I have updated the code accordingly after our chat.

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkClient.java:
```diff
@@ -0,0 +1,271 @@
+   * loss might occur.
+   */
+  public void cleanup() throws DBException {
+    synchronized (GoraBenchmarkClient.class) {
```
Can you explain the why we have chosen to use synchronization here? Should we only allow single thread to execute the close method on dataStore? Should we use isClosed similar state variable as we did for init(). If there s a possibility of getting this method cleanup() by multiple threads, anyway these threads will execute this method sequentially. Since in dataStores, we have handled the close method, even at multiple calls to close() will just try to close the already closed dataStore, it will just log and handle the error over there.

Is there any cases that we need to propagate DBException exception to higher layer? Since we dont propagate or throw exceptions at dataStore.close().

sneceesay77 replied:
I actually forgot to remove that bit of the code. It is not needed. We don't need a synchronised block in the cleanup method. So I will remove that bit of the code.

djkevincr replied:
+1

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkClient.java:
```diff
@@ -0,0 +1,271 @@
+      }
+      dataStore.put(key, user);
+    } catch (Exception e) {
+      LOG.info("There is a problem inserting data \n"+ e.getMessage(), e);
```
Please use parameterized logging here.

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkClient.java:
```diff
@@ -0,0 +1,271 @@
+  public Status insert(String table, String key, Map<String, ByteIterator> values) {
+    try {
+      user.setUserId(key);
+      for (int i = 0; i < fieldCount; i++) {
```
Please use proper variable names.

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkClient.java:
```diff
@@ -0,0 +1,271 @@
+      // Check for null is necessary.
+      User user = (fields == null || fields.size() == 0) ? dataStore.get(key)
+          : dataStore.get(key, fields.toArray(DUMMY_ARRAY));
+      for (int i = 0; i < fieldCount; i++) {
```
Please use proper variable naming.

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkClient.java:
```diff
@@ -0,0 +1,271 @@
+        result.add(hm);
+      }
+    } catch (Exception e) {
+      LOG.info("There is a problem in scanning data from the table \n"+e.getMessage(), e);
```
Please use parameterized logging here.

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkClient.java:
```diff
@@ -0,0 +1,271 @@
+      }
+      dataStore.put(user.getUserId().toString(), user);
+    } catch (Exception e) {
+      LOG.info("There is a problem updating the records \n"+e.getMessage(), e);
```
Please use parameterized logging, and avoid concatenation of Strings.

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkClient.java:
```diff
@@ -0,0 +1,271 @@
+      Result<String, User> resultSet = goraQuery.execute();
+      while (resultSet.next()) {
+        HashMap<String, ByteIterator> hm = new HashMap<>();
+        for (int i = 0; i < fieldCount; i++) {
```
Please use proper naming for variables.

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkUtils.java:
```diff
@@ -0,0 +1,428 @@
+ */
+package org.apache.gora.benchmark;
+
+import java.io.File;
```
Please add package-info.java for all source, test packages.

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkUtils.java:
```diff
@@ -0,0 +1,428 @@
+    String defaultDataStore = p.getProperty(DEFAULT_DATA_STORE_KEY);
+    String dataStore = "hbase";
+    switch (defaultDataStore) {
+    case "org.apache.gora.mongodb.store.MongoStore":
```
I think it s better to move these string constants to a separate constants file and refer all the constants from there Eg:- "org.apache.gora.mongodb.store.MongoStore", MONGODB etc.

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkUtils.java:
```diff
@@ -0,0 +1,428 @@
+   *          the p
+   * @return the data store
+   */
+  public String getDataStore(Properties p) {
```
p - Please use proper, intuitive variable naming conventions here.

djkevincr commented on gora-benchmark/src/main/resources/gora-couchdb-mapping.xml:
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Please add License header.

djkevincr commented on gora-benchmark/src/main/resources/gora-hbase-mapping.xml:
```diff
@@ -0,0 +1,59 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<gora-otd>
```
Please add License header.

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkUtils.java:
```diff
@@ -0,0 +1,428 @@
+      // setting attribute to element
+      Attr keyClass = mappingDocument.createAttribute("keyClass");
+      keyClass.setValue(KEYCLASS);
+      mappingClass.setAttributeNode(keyClass);
```
IDE shows this code as duplicated across 3 datastore ( CouchDB, MongoDB, HBase ) mapping file generation. Can we move code section to a common place?

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkUtils.java:
```diff
@@ -0,0 +1,428 @@
+      Attr tableAttribute = mappingDocument.createAttribute("name");
+      tableAttribute.setValue(jsonObject.getString("name").toLowerCase(Locale.ROOT) + "s");
+      mappingDescription.setAttributeNode(tableAttribute);
+      Element familyName = mappingDocument.createElement("family");
```
"family" - We should move these string literals to a constants file and refer from there.

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkUtils.java:
```diff
@@ -0,0 +1,428 @@
+      String jsonString = new String(Files.readAllBytes(Paths.get(fileName)), StandardCharsets.UTF_8);
+      jsonObject = new JSONObject(jsonString);
+    } catch (IOException e) {
+      LOG.info(e.getMessage(), e);
```
I think you should add proper custom error message in this case.

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkUtils.java:
```diff
@@ -0,0 +1,428 @@
+    try {
+      GoraCompiler.compileSchema(inputFiles, BEAN_DESTINATION_DIR, licenseHeader);
+    } catch (IOException e) {
+      LOG.info(e.getMessage(), e);
```
I think you should add proper custom error in this case.

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkUtils.java:
```diff
@@ -0,0 +1,428 @@
+    } catch (ParserConfigurationException e) {
+      LOG.info(e.getMessage(), e);
+    } catch (TransformerConfigurationException e) {
+      LOG.info(e.getMessage(), e);
```
I think you should add custom log messages for these cases.

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkUtils.java:
```diff
@@ -0,0 +1,428 @@
+    DocumentBuilder docBuilder;
+    try {
+      docBuilder = documentBuilderFactory.newDocumentBuilder();
+      documentBuilderFactory = DocumentBuilderFactory.newInstance();
```
documentBuilderFactory - This variable has no usage.

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkUtils.java:
```diff
@@ -0,0 +1,428 @@
+      avroWriter.write(avroSchema.toString(4));
+      avroWriter.close();
+    } catch (FileNotFoundException e) {
+      LOG.info(e.getMessage(), e);
```
Please add custom error messages for all exceptional cases.

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkUtils.java:
```diff
@@ -0,0 +1,428 @@
+      File avroFile = new File(AVRO_FULL_PATH);
+      FieldAssembler<Schema> fieldAssembler = SchemaBuilder.record(RECORD).namespace(NAMESPACE_VALUE).fields();
+      fieldAssembler.name(USER_ID_VALUE).type().stringType().stringDefault(NULL);
+      for (int i = 0; i < numberOfFields; i++) {
```
Please use proper variable names.

djkevincr commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkUtils.java:
```diff
@@ -0,0 +1,428 @@
+   *          the db name
+   */
+  @SuppressWarnings("unchecked")
+  public void generateMappingFile(String dbName) {
```
As per the current implementation of the mapping file generation, if we need to add mapping file generation to dataStore other than one's that currently support, we may need to write that logic each time. I am thinking of whether we can further generalize this generation logic further, where we maintain XSD / XML schema for different mapping files of Gora dataStores, and use that knowledge to generate mapping files. And that kind logic allows us only to write XSD for mapping files. Ideally that kind of logic we can move to gora-core.

sneceesay77 replied:
@djkevincr, this sounds good. I think @lewismc has mentioned that there is already an jira issue for such. Not sure if this will fit within the scope of the project. But in any case, I will be happy to look at this. 

Not sure about the XSD/XML schema dimension you are talking about. My current implementation uses the Avro schema as a basis for the mapping file generation.

djkevincr replied:
Yes this can be addressed separately. There is already Jira for this : https://issues.apache.org/jira/browse/GORA-455

So basically, as of now we dont have any standard way to specify a xml mapping file across all the datastores. Those xml formats are different from datastore to datastore. Make it hard to automate the generation of these file. For XML file formats we have this schema XSD, which specify a definition how your XML file will look like, what elements it will have, what attributes it will have. That XSD specification will help us to generate the XML mapping file.

sneceesay77 replied:
I will have a look at this issue after GSoC since this is the final week and there is a lot to do.

djkevincr commented on gora-benchmark/src/test/java/org/apache/gora/benchmark/GoraClientTest.java:
```diff
@@ -0,0 +1,195 @@
+  }
+
+  /**
+   * Test scan.
```
Please document these test cases.

djkevincr commented on gora-benchmark/src/test/java/org/apache/gora/benchmark/GoraClientTest.java:
```diff
@@ -0,0 +1,195 @@
+   */
+  @Test
+  public void testgenearateMappingFile() {
+    bmutils.generateMappingFile("mongodb");
```
These last 3 test cases don't have any assertion on the output which is produced by execution of the utility method.

sneceesay77 replied:
@djkevincr, don't you think I should just remove those three test? I am not very sure how to properly unit test them, any suggestions?

sneceesay77 replied:
In case, is it sufficient if the file exists and not empty?

djkevincr replied:
Yes that can be added as a basic test to check file existence. You can further add tests for content of generated xml file.
Basically you parse the xml file and check where it follows schema for that given xsd file. That is you can add tests for whether there exist a given xml element/tag after generation.

sneceesay77 replied:
I have included the basic test of checking if the file exists.

djkevincr commented on gora-benchmark/src/test/java/org/apache/gora/benchmark/GoraClientTest.java:
```diff
@@ -0,0 +1,195 @@
+  public void testGenerateDataBeans() {
+    bmutils.generateDataBeans();
+  }
+}
```
Please add new line here.

djkevincr commented:
@sneceesay77 This looks awesome and thank you for all the hard work on this. Finally I got chance to have a look your code on my IDE. I have added several comments, please resolve once you have some time. Moving forward I would like to see documentation, user guide to Gora website and documentation on analysis of methods, results you got after executing these benchmarks.

lewismc commented on gora-benchmark/src/test/java/org/apache/gora/benchmark/GoraClientTest.java:
```diff
@@ -0,0 +1,195 @@
+/**
+ * The Class GoraClientTest.
+ */
+public class GoraClientTest {
```
OK so the issue I have here is that the benchmasrk unit tests expect a local MongoDB instance ot be running... which one is not. We need to automate the provisioning of MongoDB just as is done in the [GoraMongodbTestDriver.java](https://github.com/apache/gora/blob/master/gora-mongodb/src/test/java/org/apache/gora/mongodb/GoraMongodbTestDriver.java). Please facilitate this. You may wish to just rely upon that test-jar artifact.

sneceesay77 replied:
I have updated the test to auto-provision MongoDB cluster without having to run a local copy. I have just added a simple method to this. I haven't used the test-jar or the file you have attached.

sneceesay77 commented:
@lewismc and @djkevincr, I have addressed most of the comments in this updated PR. It seems there is a lot to do in this final week of GSoC. It includes evaluations, submission of a link that hosts the project files.

@lewismc, since we are in the last week of GSoC, what options do we have if YCSB maven issue is not resolved? It seems it is not progressing well for now and I am not sure if it going to be resolved before the submission of the final evaluation.

lewismc commented on pom.xml:
```diff
@@ -826,10 +827,21 @@
     <test.container.version>1.4.2</test.container.version>
+    
+    <!-- gora-benchmark and version dependencies -->
+    <ycsb.version>0.15.0</ycsb.version>
```
This should be the newest jar version until we get the jar published to Maven Central. The user is required to download this so make it the newest.

sneceesay77 replied:
> This should be the newest jar version until we get the jar published to Maven Central. The user is required to download this so make it the newest.

@lewismc , would you please mind clarifying this for me. From https://github.com/brianfrankcooper/YCSB/releases/ it seems 0.15.0 is the latest stable release.

sneceesay77 replied:
All other comments are now addressed.

kamaci replied:
@sneceesay77 do you use a repository to include that dependency? Your version is not listed here: https://mvnrepository.com/search?q=com.yahoo.ycsb

sneceesay77 replied:
@kamaci, we have logged this issue with YCSB https://github.com/brianfrankcooper/YCSB/issues/1340, and we are currently tracking it. There is deadlock on the side of Yahoo! So for now, the workaround is to manually include the jar in the project.

kamaci replied:
Thanks for sharing the link! By the way, @lewismc pointed out 0.17.0 release to use at comments.

sneceesay77 replied:
@kamaci, thanks, I will update this accordingly.

sneceesay77 replied:
@lewismc and @kamaci, I have now updated this to 0.17.0-SNAPSHOT

kamaci commented:
Thanks for the effort you made! Could you check indentation throughout the PR files?

sneceesay77 commented:
> Thanks for the effort you made! Could you check indentation throughout the PR files?

kamaci commented on gora-benchmark/src/main/java/org/apache/gora/benchmark/GoraBenchmarkClient.java:
```diff
@@ -0,0 +1,281 @@
+/**
+ * The Class GoraBenchmarkClient
+ *
+ * @author sc306 This class extends the Yahoo! Cloud Service Benchmark DB class to
```
Who is sc306? Also, could you tidy up indentation with this Javadoc.

sneceesay77 replied:
that's my laptops username and it was auto imputed by Eclipse. I have removed all references to this now.

kamaci commented on gora-benchmark/README.md:
```diff
@@ -0,0 +1,60 @@
+<!--/**
```
No need to put asterisk characters? Check here as an example: https://raw.githubusercontent.com/kamaci/incubator-druid/feature/zookeeper_version/README.md

lewismc commented:
@sneceesay77 can you please address all updated comments and then resolve the conflict in pom.xml. For the time being also please comment out this module if it does not compile without the dependency we are waiting on. We can come back to it later on and re-enable it. Thank you

sneceesay77 commented:
> @sneceesay77 can you please address all updated comments and then resolve the conflict in pom.xml. For the time being also please comment out this module if it does not compile without the dependency we are waiting on. We can come back to it later on and re-enable it. Thank you

@lewismc and @djkevincr , I have addressed all the comments and commented out the gora-benchmark module from the list of modules. Can you please have a look and please let me know if the changes requested are satisfied?

lewismc commented:
OK @sneceesay77 we've criticized this patch pretty well... you've done an excellent job so I'm going to go ahead and merge. Good job. 
Please let us know when the YCSB dependency artifact is available and update this module with the dependency specifics at that point in time. Thank you.
